Userland/Libraries/LibTest/Randomized/Generator.h

```diff
@@ -0,0 +1,321 @@
+    });
+    bool random_bool = random_int == 1;
+    return random_bool;
+}
+
+// A (fair) boolean generator.
+//
+// Gen::boolean()
+//   -> value false, RandomRun [0]
+//   -> value true,  RandomRun [1]
+//
+// Shrinks towards false.
+inline bool boolean()
+{
+    return weighted_boolean(0.5);
+}
+
+// A vector generator of a random length between the given limits.
+//
+// Gen::vector(2,3,[]() { return Gen::unsigned_int(5); })
+//   -> value [1,5],      RandomRun [1,1,1,5,0]
+//   -> value [1,5,0],    RandomRun [1,1,1,5,1,0,0]
+//   etc.
+//
+// In case `min == max`, the RandomRun footprint will be smaller, as there will
+// be no randomness involved in figuring out the length:
+//
+// Gen::vector(3,3,[]() { return Gen::unsigned_int(5); })
+//   -> value [1,3], RandomRun [1,3]
+//   -> value [5,2], RandomRun [5,2]
+//   etc.
+//
+// Shrinks towards shorter vectors, with simpler elements inside.
+template<typename Fn>
+inline Vector<InvokeResult<Fn>> vector(size_t min, size_t max, Fn item_gen)
+{
+    VERIFY(max >= min);
+
+    size_t size = 0;
+    Vector<InvokeResult<Fn>> acc;
+
+    // Special case: no randomness for the boolean
+    if (min == max) {
+        while (size < min) {
+            acc.append(item_gen());
+            ++size;
+        }
+        return acc;
+    }
+
+    // General case: before each item we "flip a coin" to decide whether to
+    // generate another one.
+    while (size < min) {
+        acc.append(item_gen());
+        ++size;
+    }
+
+    double average = static_cast<double>(min + max) / 2.0;
+    VERIFY(average > 0);
+
+    // A geometric distribution: https://en.wikipedia.org/wiki/Geometric_distribution#Moments_and_cumulants
+    // The below derives from the E(X) = 1/p formula.
+    //
+    // We need to flip the `p` to `1-p` as our success ("another item!") is
+    // a "failure" in the geometric distribution's interpretation ("we fail X
+    // times before succeeding the first time").
+    //
+    // That gives us `1 - 1/p`. Then, E(X) also contains the final success, so we
+    // need to say `1 + average` instead of `average`, as it will mean "our X
+    // items + the final failure that stops the process".
+    double probability = 1.0 - 1.0 / (1.0 + average);
+
+    while (size < max) {
+        if (weighted_boolean(probability)) {
+            acc.append(item_gen());
+            ++size;
+        } else {
+            break;
+        }
+    }
+
+    return acc;
+}
+
+// A vector generator of a given length.
+//
+// Gen::vector_of_length(3,[]() { return Gen::unsigned_int(5); })
+//   -> value [1,5,0],    RandomRun [1,1,1,5,1,0,0]
+//   -> value [2,9,3],    RandomRun [1,2,1,9,1,3,0]
+//   etc.
+//
+// Shrinks towards shorter vectors, with simpler elements inside.
+template<typename Fn>
+inline Vector<InvokeResult<Fn>> vector(size_t length, Fn item_gen)
+{
+    return vector(length, length, item_gen);
+}
+
+// A vector generator of a random length between 0 and 32 elements.
```

Comment:
Was there any notable thought behind the upper limit in particular? 32 seems a bit excessive to me, and I'd probably rather want to force the test writer to think of a reasonable limit anyways.

Reply:
Not really, it's just a reasonable default.

There are some anecdotes that some interesting bugs appear after eg. tens or hundreds of randomly generated steps rather than let's say <10:

- [Google LevelDB](https://htmlpreview.github.io/?https://raw.github.com/strangeloop/lambdajam2013/master/slides/Norton-QuickCheck.html#21.0)
  - TL;DR the test generated 17 steps (well, generated more then shrunk those to 17!) that triggered a bug, then Google fixed it, then the same test found 33 steps for the same bug
- [Quviq testing Ericsson's media proxy system](http://www.quviq.com/wp-content/uploads/2014/08/erlang001-arts.pdf) - section 5.1, last bulletpoint
  - TL;DR the test generated around 160 steps that triggered a bug and shrunk that list down to 7 steps

That is to say, I think there's value in having a chance to occasionally generate larger vectors when using the default "just give me a vector" API.

I'd like to keep the `vector(item)` signature (rather than forcing people to always give a length / a length range) -- my view is that it's a small miracle that people test their code at all, so I try to lower the barriers to writing the test easily. But, if you think we'll be better off with user always picking a length range, we can remove that specific signature!

Reply:
Sounds good.